File: src/Python/brlcad/VectorList.py
#                       V E C T O R L I S T . P Y
#  BRL-CAD
#
# Copyright (c) 2026 United States Government as represented by
# the U.S. Army Research Laboratory.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# version 2.1 as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this file; see the file named COPYING for more
# information.
#
# @file VectorList.py
#
# BRL-CAD core simplified Python interface:
#       Python interface implementation for VectorList.cpp

from ._bindings import _lib, BrlVectorListCallback
from .Handle import Handle


class ElementType:
    PointDraw = 0
    PointSize = 1
    LineMove = 2
    LineDraw = 3
    LineWidth = 4
    TriangleStart = 5
    TriangleMove = 6
    TriangleDraw = 7
    TriangleEnd = 8
    TriangleVertexNormal = 9
    PolygonStart = 10
    PolygonMove = 11
    PolygonDraw = 12
    PolygonEnd = 13
    PolygonVertexNormal = 14
    DisplaySpace = 15
    ModelSpace = 16


def _vector3d_tuple(vec_handle):
    if not vec_handle:
        return None
    x = _lib.BrlVector3DX(vec_handle)
    y = _lib.BrlVector3DY(vec_handle)
    z = _lib.BrlVector3DZ(vec_handle)
    _lib.BrlDeleteHandle(vec_handle)
    return (x, y, z)


class VectorListElement(Handle):
    """Borrowed or owned VectorList element handle."""

    def __init__(self, handle, owned=True):
        super().__init__(handle=handle, owned=owned)

    @property
    def element_type(self):
        if not self._handle:
            return None
        return _lib.BrlVectorListElementGetType(self._handle)


class VectorListPointDraw(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListPointDraw(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListPointDrawPoint(self._handle))


class VectorListPointSize(VectorListElement):
    def __init__(self, size=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            super().__init__(handle=_lib.BrlNewVectorListPointSize(size), owned=True)

    @property
    def size(self):
        return _lib.BrlVectorListPointSizeSize(self._handle)


class VectorListLineMove(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListLineMove(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListLineMovePoint(self._handle))


class VectorListLineDraw(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListLineDraw(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListLineDrawPoint(self._handle))


class VectorListLineWidth(VectorListElement):
    def __init__(self, width=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            super().__init__(handle=_lib.BrlNewVectorListLineWidth(width), owned=True)

    @property
    def width(self):
        return _lib.BrlVectorListLineWidthWidth(self._handle)


class VectorListTriangleStart(VectorListElement):
    def __init__(self, normal=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            nx, ny, nz = normal
            super().__init__(handle=_lib.BrlNewVectorListTriangleStart(nx, ny, nz), owned=True)

    @property
    def normal(self):
        return _vector3d_tuple(_lib.BrlVectorListTriangleStartNormal(self._handle))


class VectorListTriangleMove(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListTriangleMove(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListTriangleMovePoint(self._handle))


class VectorListTriangleDraw(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListTriangleDraw(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListTriangleDrawPoint(self._handle))


class VectorListTriangleEnd(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListTriangleEnd(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListTriangleEndPoint(self._handle))


class VectorListTriangleVertexNormal(VectorListElement):
    def __init__(self, normal=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            nx, ny, nz = normal
            super().__init__(handle=_lib.BrlNewVectorListTriangleVertexNormal(nx, ny, nz), owned=True)

    @property
    def normal(self):
        return _vector3d_tuple(_lib.BrlVectorListTriangleVertexNormalNormal(self._handle))


class VectorListPolygonStart(VectorListElement):
    def __init__(self, normal=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            nx, ny, nz = normal
            super().__init__(handle=_lib.BrlNewVectorListPolygonStart(nx, ny, nz), owned=True)

    @property
    def normal(self):
        return _vector3d_tuple(_lib.BrlVectorListPolygonStartNormal(self._handle))


class VectorListPolygonMove(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListPolygonMove(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListPolygonMovePoint(self._handle))


class VectorListPolygonDraw(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListPolygonDraw(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListPolygonDrawPoint(self._handle))


class VectorListPolygonEnd(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListPolygonEnd(x, y, z), owned=True)

    @property
    def point(self):
        return _vector3d_tuple(_lib.BrlVectorListPolygonEndPoint(self._handle))


class VectorListPolygonVertexNormal(VectorListElement):
    def __init__(self, normal=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            nx, ny, nz = normal
            super().__init__(handle=_lib.BrlNewVectorListPolygonVertexNormal(nx, ny, nz), owned=True)

    @property
    def normal(self):
        return _vector3d_tuple(_lib.BrlVectorListPolygonVertexNormalNormal(self._handle))


class VectorListDisplaySpace(VectorListElement):
    def __init__(self, point=None, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            x, y, z = point
            super().__init__(handle=_lib.BrlNewVectorListDisplaySpace(x, y, z), owned=True)

    @property
    def reference_point(self):
        return _vector3d_tuple(_lib.BrlVectorListDisplaySpaceReferencePoint(self._handle))


class VectorListModelSpace(VectorListElement):
    def __init__(self, handle=None, owned=True):
        if handle is not None:
            super().__init__(handle=handle, owned=owned)
        else:
            super().__init__(handle=_lib.BrlNewVectorListModelSpace(), owned=True)


_ELEMENT_CLASSES = {
    ElementType.PointDraw: VectorListPointDraw,
    ElementType.PointSize: VectorListPointSize,
    ElementType.LineMove: VectorListLineMove,
    ElementType.LineDraw: VectorListLineDraw,
    ElementType.LineWidth: VectorListLineWidth,
    ElementType.TriangleStart: VectorListTriangleStart,
    ElementType.TriangleMove: VectorListTriangleMove,
    ElementType.TriangleDraw: VectorListTriangleDraw,
    ElementType.TriangleEnd: VectorListTriangleEnd,
    ElementType.TriangleVertexNormal: VectorListTriangleVertexNormal,
    ElementType.PolygonStart: VectorListPolygonStart,
    ElementType.PolygonMove: VectorListPolygonMove,
    ElementType.PolygonDraw: VectorListPolygonDraw,
    ElementType.PolygonEnd: VectorListPolygonEnd,
    ElementType.PolygonVertexNormal: VectorListPolygonVertexNormal,
    ElementType.DisplaySpace: VectorListDisplaySpace,
    ElementType.ModelSpace: VectorListModelSpace,
}


def _wrap_borrowed_element(handle):
    if not handle:
        return None
    element_type = _lib.BrlVectorListElementGetType(handle)
    cls = _ELEMENT_CLASSES.get(element_type, VectorListElement)
    return cls(handle=handle, owned=False)


class VectorList(Handle):
    """
    Object-oriented Python interface for the BRL-CAD VectorList.
    Typically filled by Plot, then read via iterate().
    """

    def __init__(self):
        native_handle = _lib.BrlNewVectorList()
        if not native_handle:
            native_handle = None
        super().__init__(handle=native_handle)

    def clear(self):
        if not self._handle:
            return
        _lib.BrlVectorListClear(self._handle)

    def iterate(self, callback):
        """
        Call callback(element) for each element.
        Return False from callback to stop; True/None continues.
        Element handles are borrowed and only valid during the callback.
        """
        if not self._handle:
            return

        @BrlVectorListCallback
        def c_callback(element_handle, userdata):
            element = _wrap_borrowed_element(element_handle)
            result = callback(element)
            if result is False:
                return 0
            return 1

        _lib.BrlVectorListIterate(self._handle, c_callback, None)

    def elements(self):
        """Collect a snapshot of readable element data during iteration."""
        items = []

        def collect(element):
            entry = {'type': element.element_type}
            if hasattr(element, 'point'):
                entry['point'] = element.point
            if hasattr(element, 'normal'):
                entry['normal'] = element.normal
            if hasattr(element, 'size'):
                entry['size'] = element.size
            if hasattr(element, 'width'):
                entry['width'] = element.width
            if hasattr(element, 'reference_point'):
                entry['reference_point'] = element.reference_point
            items.append(entry)
            return True

        self.iterate(collect)
        return items

    def append(self, element):
        """Secondary write helper."""
        if not self._handle or not element._handle:
            return False
        return _lib.BrlVectorListAppend(self._handle, element._handle) != 0

File: src/C/vectorList.cpp
/*                      V E C T O R L I S T . C P P
 * BRL-CAD
 *
 * Copyright (c) 2026 United States Government as represented by
 * the U.S. Army Research Laboratory.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this file; see the file named COPYING for more
 * information.
 */
/** @file vectorList.cpp
 *
 *  BRL-CAD core simplified C interface:
 *      implementation for VectorList opaque wrapper
 */

#include <cassert>
#include <functional>

#include <brlcad/VectorList.h>

#include <brlcad/C/vectorList.h>

#include "casts.h"


using namespace BRLCAD;


namespace {


VectorList::Element* ElementOrNull
(
    BrlVectorListElement element
) {
    VectorList::Element* ret = nullptr;

    if (element != nullptr) {
        ret = CastVectorListElement(element);
        assert(ret != nullptr);
    }

    return ret;
}


BrlVectorListElement CastIfType
(
    BrlVectorListElement           element,
    VectorList::Element::ElementType type
) {
    BrlVectorListElement ret = nullptr;
    VectorList::Element* el  = ElementOrNull(element);

    if (el != nullptr && el->Type() == type)
        ret = element;

    return ret;
}


BrlVector3D PointIfType
(
    BrlVectorListElement           element,
    VectorList::Element::ElementType type
) {
    BrlVector3D          ret = nullptr;
    VectorList::Element* el  = ElementOrNull(element);

    if (el != nullptr && el->Type() == type) {
        switch (type) {
            case VectorList::Element::ElementType::PointDraw:
                ret = new Vector3DData(static_cast<VectorList::PointDraw*>(el)->Point());
                break;
            case VectorList::Element::ElementType::LineMove:
                ret = new Vector3DData(static_cast<VectorList::LineMove*>(el)->Point());
                break;
            case VectorList::Element::ElementType::LineDraw:
                ret = new Vector3DData(static_cast<VectorList::LineDraw*>(el)->Point());
                break;
            case VectorList::Element::ElementType::TriangleMove:
                ret = new Vector3DData(static_cast<VectorList::TriangleMove*>(el)->Point());
                break;
            case VectorList::Element::ElementType::TriangleDraw:
                ret = new Vector3DData(static_cast<VectorList::TriangleDraw*>(el)->Point());
                break;
            case VectorList::Element::ElementType::TriangleEnd:
                ret = new Vector3DData(static_cast<VectorList::TriangleEnd*>(el)->Point());
                break;
            case VectorList::Element::ElementType::PolygonMove:
                ret = new Vector3DData(static_cast<VectorList::PolygonMove*>(el)->Point());
                break;
            case VectorList::Element::ElementType::PolygonDraw:
                ret = new Vector3DData(static_cast<VectorList::PolygonDraw*>(el)->Point());
                break;
            case VectorList::Element::ElementType::PolygonEnd:
                ret = new Vector3DData(static_cast<VectorList::PolygonEnd*>(el)->Point());
                break;
            default:
                break;
        }
    }

    return ret;
}


BrlVector3D NormalIfType
(
    BrlVectorListElement           element,
    VectorList::Element::ElementType type
) {
    BrlVector3D          ret = nullptr;
    VectorList::Element* el  = ElementOrNull(element);

    if (el != nullptr && el->Type() == type) {
        switch (type) {
            case VectorList::Element::ElementType::TriangleStart:
                ret = new Vector3DData(static_cast<VectorList::TriangleStart*>(el)->Normal());
                break;
            case VectorList::Element::ElementType::TriangleVertexNormal:
                ret = new Vector3DData(static_cast<VectorList::TriangleVertexNormal*>(el)->Normal());
                break;
            case VectorList::Element::ElementType::PolygonStart:
                ret = new Vector3DData(static_cast<VectorList::PolygonStart*>(el)->Normal());
                break;
            case VectorList::Element::ElementType::PolygonVertexNormal:
                ret = new Vector3DData(static_cast<VectorList::PolygonVertexNormal*>(el)->Normal());
                break;
            default:
                break;
        }
    }

    return ret;
}


}


BrlVectorList BrlNewVectorList(void) {
    return new VectorListData(new VectorList());
}


void BrlVectorListClear
(
    BrlVectorList vlist
) {
    if (vlist != nullptr) {
        VectorList* vl = CastVectorList(vlist);

        assert(vl != nullptr);

        if (vl != nullptr)
            vl->Clear();
    }
}


void BrlVectorListIterate
(
    BrlVectorList         vlist,
    BrlVectorListCallback callback,
    void*                 userdata
) {
    if (vlist != nullptr && callback != nullptr) {
        VectorList* vl = CastVectorList(vlist);

        assert(vl != nullptr);

        if (vl != nullptr) {
            vl->Iterate([callback, userdata](const VectorList::Element* element) -> bool {
                if (element == nullptr)
                    return true;

                BrlVectorListElement handle = new VectorListElementData(
                    const_cast<VectorList::Element*>(element), false);
                int cont = callback(handle, userdata);

                delete static_cast<VectorListElementData*>(handle);

                return cont != 0;
            });
        }
    }
}


BrlVectorListElementType BrlVectorListElementGetType
(
    BrlVectorListElement element
) {
    BrlVectorListElementType ret = BrlVectorListElementTypePointDraw;
    VectorList::Element*     el  = ElementOrNull(element);

    if (el != nullptr)
        ret = static_cast<BrlVectorListElementType>(el->Type());

    return ret;
}


BrlVectorListElement BrlCastToVectorListPointDraw(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PointDraw);
}

BrlVectorListElement BrlCastToVectorListPointSize(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PointSize);
}

BrlVectorListElement BrlCastToVectorListLineMove(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::LineMove);
}

BrlVectorListElement BrlCastToVectorListLineDraw(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::LineDraw);
}

BrlVectorListElement BrlCastToVectorListLineWidth(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::LineWidth);
}

BrlVectorListElement BrlCastToVectorListTriangleStart(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::TriangleStart);
}

BrlVectorListElement BrlCastToVectorListTriangleMove(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::TriangleMove);
}

BrlVectorListElement BrlCastToVectorListTriangleDraw(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::TriangleDraw);
}

BrlVectorListElement BrlCastToVectorListTriangleEnd(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::TriangleEnd);
}

BrlVectorListElement BrlCastToVectorListTriangleVertexNormal(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::TriangleVertexNormal);
}

BrlVectorListElement BrlCastToVectorListPolygonStart(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PolygonStart);
}

BrlVectorListElement BrlCastToVectorListPolygonMove(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PolygonMove);
}

BrlVectorListElement BrlCastToVectorListPolygonDraw(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PolygonDraw);
}

BrlVectorListElement BrlCastToVectorListPolygonEnd(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PolygonEnd);
}

BrlVectorListElement BrlCastToVectorListPolygonVertexNormal(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::PolygonVertexNormal);
}

BrlVectorListElement BrlCastToVectorListDisplaySpace(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::DisplaySpace);
}

BrlVectorListElement BrlCastToVectorListModelSpace(BrlVectorListElement element) {
    return CastIfType(element, VectorList::Element::ElementType::ModelSpace);
}


BrlVector3D BrlVectorListPointDrawPoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::PointDraw);
}


double BrlVectorListPointSizeSize(BrlVectorListElement element) {
    double               ret = 0.;
    VectorList::Element* el  = ElementOrNull(element);

    if (el != nullptr && el->Type() == VectorList::Element::ElementType::PointSize)
        ret = static_cast<VectorList::PointSize*>(el)->Size();

    return ret;
}


BrlVector3D BrlVectorListLineMovePoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::LineMove);
}


BrlVector3D BrlVectorListLineDrawPoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::LineDraw);
}


double BrlVectorListLineWidthWidth(BrlVectorListElement element) {
    double               ret = 0.;
    VectorList::Element* el  = ElementOrNull(element);

    if (el != nullptr && el->Type() == VectorList::Element::ElementType::LineWidth)
        ret = static_cast<VectorList::LineWidth*>(el)->Width();

    return ret;
}


BrlVector3D BrlVectorListTriangleStartNormal(BrlVectorListElement element) {
    return NormalIfType(element, VectorList::Element::ElementType::TriangleStart);
}


BrlVector3D BrlVectorListTriangleMovePoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::TriangleMove);
}


BrlVector3D BrlVectorListTriangleDrawPoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::TriangleDraw);
}


BrlVector3D BrlVectorListTriangleEndPoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::TriangleEnd);
}


BrlVector3D BrlVectorListTriangleVertexNormalNormal(BrlVectorListElement element) {
    return NormalIfType(element, VectorList::Element::ElementType::TriangleVertexNormal);
}


BrlVector3D BrlVectorListPolygonStartNormal(BrlVectorListElement element) {
    return NormalIfType(element, VectorList::Element::ElementType::PolygonStart);
}


BrlVector3D BrlVectorListPolygonMovePoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::PolygonMove);
}


BrlVector3D BrlVectorListPolygonDrawPoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::PolygonDraw);
}


BrlVector3D BrlVectorListPolygonEndPoint(BrlVectorListElement element) {
    return PointIfType(element, VectorList::Element::ElementType::PolygonEnd);
}


BrlVector3D BrlVectorListPolygonVertexNormalNormal(BrlVectorListElement element) {
    return NormalIfType(element, VectorList::Element::ElementType::PolygonVertexNormal);
}


BrlVector3D BrlVectorListDisplaySpaceReferencePoint(BrlVectorListElement element) {
    BrlVector3D          ret = nullptr;
    VectorList::Element* el  = ElementOrNull(element);

    if (el != nullptr && el->Type() == VectorList::Element::ElementType::DisplaySpace)
        ret = new Vector3DData(static_cast<VectorList::DisplaySpace*>(el)->ReferencePoint());

    return ret;
}


int BrlVectorListAppend
(
    BrlVectorList        vlist,
    BrlVectorListElement element
) {
    int ret = 0;

    if (vlist != nullptr && element != nullptr) {
        VectorList*          vl = CastVectorList(vlist);
        VectorList::Element* el = CastVectorListElement(element);

        assert(vl != nullptr && el != nullptr);

        if (vl != nullptr && el != nullptr)
            ret = vl->Append(*el) ? 1 : 0;
    }

    return ret;
}


BrlVectorListElement BrlNewVectorListPointDraw
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::PointDraw(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListPointSize
(
    double size
) {
    return new VectorListElementData(new VectorList::PointSize(size));
}


BrlVectorListElement BrlNewVectorListLineMove
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::LineMove(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListLineDraw
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::LineDraw(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListLineWidth
(
    double width
) {
    return new VectorListElementData(new VectorList::LineWidth(width));
}


BrlVectorListElement BrlNewVectorListTriangleStart
(
    double nx,
    double ny,
    double nz
) {
    return new VectorListElementData(new VectorList::TriangleStart(Vector3D(nx, ny, nz)));
}


BrlVectorListElement BrlNewVectorListTriangleMove
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::TriangleMove(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListTriangleDraw
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::TriangleDraw(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListTriangleEnd
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::TriangleEnd(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListTriangleVertexNormal
(
    double nx,
    double ny,
    double nz
) {
    return new VectorListElementData(new VectorList::TriangleVertexNormal(Vector3D(nx, ny, nz)));
}


BrlVectorListElement BrlNewVectorListPolygonStart
(
    double nx,
    double ny,
    double nz
) {
    return new VectorListElementData(new VectorList::PolygonStart(Vector3D(nx, ny, nz)));
}


BrlVectorListElement BrlNewVectorListPolygonMove
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::PolygonMove(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListPolygonDraw
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::PolygonDraw(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListPolygonEnd
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::PolygonEnd(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListPolygonVertexNormal
(
    double nx,
    double ny,
    double nz
) {
    return new VectorListElementData(new VectorList::PolygonVertexNormal(Vector3D(nx, ny, nz)));
}


BrlVectorListElement BrlNewVectorListDisplaySpace
(
    double x,
    double y,
    double z
) {
    return new VectorListElementData(new VectorList::DisplaySpace(Vector3D(x, y, z)));
}


BrlVectorListElement BrlNewVectorListModelSpace(void) {
    return new VectorListElementData(new VectorList::ModelSpace());
}
